drivers/cache/cache_andes.c

```diff
@@ -0,0 +1,737 @@
+	if (csr_read(NDS_MCACHE_CTL) & BIT(20)) {
+		csr_clear(NDS_MCACHE_CTL, BIT(19));
+		/* Wait for cache coherence disabling completed */
+		while (csr_read(NDS_MCACHE_CTL) & BIT(20)) {
+			;
+		}
+	}
+
+#ifdef L2C_EXISTS
+	if (cache_cfg.l2c_size) {
+		uint32_t l2c_ctrl = sys_read32(L2C_CTRL);
+
+		if (l2c_ctrl & BIT(0)) {
+			l2c_ctrl &= ~BIT(0);
+			sys_write32(l2c_ctrl, L2C_CTRL);
+
+			uint8_t hart_id = arch_proc_id();
+
+			sys_write32(CCTL_L2_WBINVAL_ALL, L2C_CCTLCMD(hart_id));
+
+			/* Wait L2 CCTL Commands finished */
+			nds_l2_cache_wait_status(hart_id);
+		}
+	}
+#endif /* L2C_EXISTS */
+
+	k_spin_unlock(&lock, key);
+}
+
+void cache_instr_enable(void)
+{
+#if defined(CONFIG_SMP) && (CONFIG_MP_MAX_NUM_CPUS > 1)
+	return;
+#endif
+	csr_set(NDS_MCACHE_CTL, BIT(0));
+}
+
+void cache_instr_disable(void)
+{
+#if defined(CONFIG_SMP) && (CONFIG_MP_MAX_NUM_CPUS > 1)
+	return;
+#endif
+	csr_clear(NDS_MCACHE_CTL, BIT(0));
+}
+
+int cache_data_invd_all(void)
+{
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+#ifdef CONFIG_L2C_INCLUSIVE_POLICY
+	if (cache_cfg.l2c_version > 15) {
+		nds_l2_cache_all(K_CACHE_WB);
+		nds_l2_cache_all(K_CACHE_INVD);
+	} else {
+		nds_l1d_cache_all(K_CACHE_WB);
+		nds_l2_cache_all(K_CACHE_WB);
+
+		nds_l2_cache_all(K_CACHE_INVD);
+		nds_l1d_cache_all(K_CACHE_INVD);
+	}
+#else
+	nds_l1d_cache_all(K_CACHE_WB);
+	nds_l2_cache_all(K_CACHE_WB);
+
+	nds_l2_cache_all(K_CACHE_INVD);
+	nds_l1d_cache_all(K_CACHE_INVD);
+#endif /* CONFIG_L2C_INCLUSIVE_POLICY */
+
+	k_spin_unlock(&lock, key);
+
+	return 0;
+}
+
+int cache_data_invd_range(void *addr, size_t size)
+{
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+#ifdef CONFIG_L2C_INCLUSIVE_POLICY
+	if (cache_cfg.l2c_version > 15) {
+		nds_l2_cache_range(addr, size, K_CACHE_INVD);
+	} else {
+		nds_l2_cache_range(addr, size, K_CACHE_INVD);
+		nds_l1d_cache_range(addr, size, K_CACHE_INVD);
+	}
+#else
+	nds_l2_cache_range(addr, size, K_CACHE_INVD);
+	nds_l1d_cache_range(addr, size, K_CACHE_INVD);
+#endif /* CONFIG_L2C_INCLUSIVE_POLICY */
+
+	k_spin_unlock(&lock, key);
+
+	return 0;
+}
+
+int cache_instr_invd_all(void)
+{
+#if defined(CONFIG_SMP) && (CONFIG_MP_MAX_NUM_CPUS > 1)
+	return -ENOTSUP;
+#endif
+
+#ifdef CONFIG_RISCV_PMP
+	/*  CCTL IX type command is not to RISC-V Smepmp */
+#if CONFIG_64BIT
+	if (csr_read(NDS_MMSC_CFG) & BIT(52)) {
+#else
+	if ((csr_read(NDS_MMSC_CFG) & BIT(31)) &&
+		(csr_read(NDS_MMSC_CFG2) & BIT(20))) {
+#endif /* CONFIG_64BIT */
+		return -ENOTSUP;
+	}
+#endif /* CONFIG_RISCV_PMP */
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+	nds_l1i_cache_all(K_CACHE_INVD);
+
+	k_spin_unlock(&lock, key);
+
+	return 0;
+}
+
+int cache_instr_invd_range(void *addr, size_t size)
+{
+#if defined(CONFIG_SMP) && (CONFIG_MP_MAX_NUM_CPUS > 1)
+	ARG_UNUSED(addr);
+	ARG_UNUSED(size);
+
+	return -ENOTSUP;
+#endif
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+	nds_l1i_cache_range(addr, size, K_CACHE_INVD);
+
+	k_spin_unlock(&lock, key);
```

Comment:
this and many others can use some readability optimizations:

```suggestion
	K_SPINLOCK(&lock) {
		nds_l1i_cache_all(K_CACHE_INVD);
	}
```